opentelemetry-api/src/opentelemetry/context/base_context.py

```diff
@@ -0,0 +1,116 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import threading
+import typing
+
+
+def wrap_callable(target: 'object') -> typing.Callable[[], object]:
+    if callable(target):
+        return target
+    return lambda: target
+
+
+class BaseRuntimeContext:
+    class Slot:
+        def __init__(self, name: str, default: 'object'):
+            raise NotImplementedError
+
+        def clear(self) -> None:
+            raise NotImplementedError
+
+        def get(self) -> 'object':
+            raise NotImplementedError
+
+        def set(self, value: 'object') -> None:
+            raise NotImplementedError
+
+    _lock = threading.Lock()
+    _slots: typing.Dict[str, Slot] = {}
+
+    @classmethod
+    def clear(cls) -> None:
+        """Clear all slots to their default value."""
+        keys = cls._slots.keys()
+        for name in keys:
+            slot = cls._slots[name]
+            slot.clear()
+
+    @classmethod
+    def register_slot(cls, name: str, default: 'object' = None) -> 'Slot':
+        """Register a context slot with an optional default value.
+
+        :type name: str
+        :param name: The name of the context slot.
+
+        :type default: object
+        :param name: The default value of the slot, can be a value or lambda.
+
+        :returns: The registered slot.
+        """
+        with cls._lock:
+            if name in cls._slots:
+                raise ValueError('slot {} already registered'.format(name))
+            slot = cls.Slot(name, default)
+            cls._slots[name] = slot
+            return slot
+
+    def apply(self, snapshot: typing.Dict[str, 'object']) -> None:
+        """Set the current context from a given snapshot dictionary"""
+
+        for name in snapshot:
+            setattr(self, name, snapshot[name])
+
+    def snapshot(self) -> typing.Dict[str, 'object']:
+        """Return a dictionary of current slots by reference."""
+
+        keys = self._slots.keys()
+        return dict((n, self._slots[n].get()) for n in keys)
+
+    def __repr__(self) -> str:
+        return '{}({})'.format(type(self).__name__, self.snapshot())
+
+    def __getattr__(self, name: str) -> 'object':
+        if name not in self._slots:
+            self.register_slot(name, None)
+        slot = self._slots[name]
+        return slot.get()
+
+    def __setattr__(self, name: str, value: 'object') -> None:
+        if name not in self._slots:
+            self.register_slot(name, None)
+        slot = self._slots[name]
+        slot.set(value)
+
+    def with_current_context(
```

Comment:
would there be value in exposing a contextmanager like this as well?

Reply:
I don't know at this moment. I'm open to add it if we find it helpful.